### DISCUSSION TITLE
Schema from code questions

### DISCUSSION THREAD

**Discussion by brmakana** [Netflix/dgs-framework#206] [Q&A]
Hello,
I have a graphql-java application with a lot of custom code around translating my domain model to GraphQL types, and I'm interested in using DGS and the advanced feature "schema from code", but running into some stumbling blocks. 

My code handles creating `GraphQLOutputType`s and `GraphQLInputType`s from my POJOs, but it looks like the DGS way revolves around working with a `TypeDefinitionRegistry`. 

I'm wondering if anyone can provide some pointers on how to go from `GraphQLOutputType`/`GraphQLInputType` to something `TypeDefinitionRegistry` can understand? `GraphQLObjectType` does have a `.getDefinition()`, but TypeDefinitionRegistry takes an `ObjectTypeExtensionDefinition`, not an `ObjectTypeDefinition`. As well, I don't see any way to convert a GraphQLList..

Any pointers are welcome, I think DGS looks really exciting!

**Comment by srinivasankavitha** [comment]
Yeah, unfortunately, we don't currently have a way of adding ObjectType to the TypeDefinitionRegistry. We will need to add support for that.

**Reply by brmakana** [reply]
Thanks for the reply @srinivasankavitha. I can construct ObjectTypeDefinitions programmatically, as it looks similar to how you construct GraphQLObject types; however, I'm still not sure how to build the ObjectTypeDefinition representation of a GraphQLList (or whatever SDLDefinition that GraphQLList is)

**Reply by srinivasankavitha** [reply]
Have you tried using `graphql.schema.GraphQLList.list`?

```
String::class.java  -> list(Scalars.GraphQLString)
Integer::class.java -> list(Scalars.GraphQLInt)
Boolean::class.java -> list(Scalars.GraphQLBoolean)
Float::class.java -> list(Scalars.GraphQLFloat)
or
ListType(TypeName(nameOfType))

```

**Reply by srinivasankavitha** [reply]
Note that the above is to build an ObjectType, not an ObjectTypeExtensionDefinition, assuming the former is what you are asking about. If you want to build a list that is compatible with ObjectTypeDefinition and ObjectTypeExtensionDefinition, you can try using `graphql.language.ListType`

```
String::class.java  -> ListType(TypeName("String"))
Integer::class.java -> ListType(TypeName("Int"))
Boolean::class.java -> ListType(TypeName("Boolean"))
Float::class.java -> ListType(TypeName("Float"))
or
ListType(TypeName(nameOfType))
```

**Reply by brmakana** [reply]
I think I am more confused now.. previously with graphql-java, to create my schema programmatically looked like this: 
```
              final GraphQLObjectType cat = GraphQLObjectType.newObject()
                .name("Cat")
                .field(GraphQLFieldDefinition.newFieldDefinition()
                    .name("name")
                    .type(GraphQLString)
                    .build())
                .build();
        final GraphQLObjectType query = GraphQLObjectType.newObject()
            .name("QueryType")
            .field(GraphQLFieldDefinition.newFieldDefinition()
                .name("getCats")
                .type(cat)
                )
            .build();
            
        final GraphQLSchema schema = GraphQLSchema.newSchema()
//            .codeRegistry() ignore datafetcher for this example
            .query(query)
            .build();            
```
with DGS, I can make a TypeDefinitionRegistry, but I'm confused on how to do the equivalent: 
```
    @DgsTypeDefinitionRegistry
    public TypeDefinitionRegistry registry() {
        final TypeDefinitionRegistry typeDefinitionRegistry = new TypeDefinitionRegistry();
        final ObjectTypeExtensionDefinition cat = ObjectTypeExtensionDefinition.newObjectTypeExtensionDefinition()
            .name("Cat")
            .fieldDefinition(FieldDefinition.newFieldDefinition()
                .name("name")
                .type(new TypeName("String"))
                .build()
            )
            .build();
        typeDefinitionRegistry.add(cat);
        return typeDefinitionRegistry;
    }
```
this results in: 
```
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [graphql.schema.GraphQLSchema]: Factory method 'schema' threw exception; nested exception is SchemaProblem{errors=[The extension 'Cat' type [@-1:-1] is missing its base underlying type]}
```

I appreciate your help!

**Reply by srinivasankavitha** [reply]
Can you try changing it to ObjectTypeDefinition instead of Extension? For everything to work together, you will need to add the Query type as well with the corresponding field. This should give you the schema shown below:
```
type Cat {
    name: String
}
```

```
@DgsTypeDefinitionRegistry
    public TypeDefinitionRegistry registry() {
        final TypeDefinitionRegistry typeDefinitionRegistry = new TypeDefinitionRegistry();
        final ObjectTypeExtensionDefinition cat = ObjectTypeDefinition.newObjectTypeDefinition()
            .name("Cat")
            .fieldDefinition(FieldDefinition.newFieldDefinition()
                .name("name")
                .type(new TypeName("String"))
                .build()
            )
            .build();
        typeDefinitionRegistry.add(cat);
        return typeDefinitionRegistry;
    }

```

For the Query type, you can then add an `ObjectTypeExtensionDefinition` to extend the base query type.
```
type Query extends {
    cat: Cat
}

ObjectTypeExtensionDefinition query = ObjectTypeExtensionDefinition.newObjectTypeExtensionDefinition().name("Query").fieldDefinition(
                FieldDefinition.newFieldDefinition().name("cat").type(new TypeName("Cat")).build()
        ).build();
```

We'll add more clarification around this in the docs so it's less confusing. Hope this helps.

**Reply by brmakana** [reply]
Ah, that worked! Thank you so much for the help!
